<h2>FIN5615 Project #1
    <p class="text-primary">Time Value of Money</p>
</h2>

<div class = "alert alert-block alert-danger">
    <b>Due Date</b>: October 5th, 2020, 11:59PM EST
</div>  

<div class = "panel panel-success">
    <div class = "panel-heading"> 
    <span style="color:black"><b>Project Introduction</b>
    </div>
    <div class = "panel-body"> 
    In this project we will examine the time value of money and the mechanics of discounting future cashflows to present day. We will then attempt to apply these concepts to valuing bonds and equities.
    </div>
</div>

<div class = "panel panel-info">
    <div class = "panel-heading"> 
    <span style="color:black"><b>Instructions</b>
    </div>
    <div class = "panel-body"> 
    Please code you solutions directly in this Notebook under each task in the cells provided below. Include all code used to produce your results, and support your answers with textual explanations if neccessary. You may feel free to add additional Code and/or Markdown cells as needed. Code must be written using <b>Python 3</b>.
    <br><br>Please order the cells in your Notebook in the order in which they are intended to run! It is expected that your Notebook is able to run in sequence from start to finish without any errors (before submitting check by: <code>Kernel > Restart & Run All</code>).
    <br><br>Once ready, please submit your assignment as an <b>.ipynb file</b> (<code>File > Download as > Notebook</code>). You must name your files in the following format: <b><code>FIN5615_Assignmnet_1_Richard_Lee_tuc12345.ipynb</code></b>.
    </div>
</div>

---
<div class = "alert alert-block alert-warning">
    <b>Note</b>: Import all libraries that you plan on using in the code block below.
</div>

In [1]:
#Nick Irrer
import numpy as np

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 1</b>
    </div>
    <div class = "panel-body"> 
    Time value of money is the concept that receiving a cashflow today is worth more than receiving the same amount at a future date. We can compare cashflows of differing time values by calculating and comparing their present values.
    <br><br>Recall that the formula to calculate the present value of a cashflow <i>CF</i> at time <i>t</i> assuming annual compounding at rate <i>r</i> is: 
    <br><br>
    $$PV = \frac{CF}{(1 + r)^t}$$
    <br>Compute the present value of a cashflow based on the inputs below.
    <p style="text-align:right;">[5pts]</p>
    </div>
</div>

In [2]:
CF = 100.
rate = .05
t = 2

#_#Your code here

present_value=CF/(1+rate)**t
print(present_value)

#_#Computing the present value with given cashflows, rate, and time

90.70294784580499


<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 2</b>
    </div>
    <div class = "panel-body"> 
    The present value formula can be modified to account for periodic compounding at some m-thly frequency:
    <br><br>
    $$PV = \frac{CF}{\left(1 + \frac{r}{m}\right)^{mt}}$$
    <br>Assuming the same inputs as the previous task, now include a new variable *m* with a value of 12 representing monthly compounding. Calculate the cashflow's present value given monthly compounding.
    <p style="text-align:right;">[10pts]</p>
    </div>
</div>

In [3]:
m = 12

present_value_monthly_compounding=CF/(1+rate/m)**(m*t)
print(present_value_monthly_compounding)

#_#Computing present value with monthly compounding using same CF, r, t.


90.5025423358517


<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 3</b>
    </div>
    <div class = "panel-body"> 
    Calculate the cashflow's present value under daily compounding.
    <br><br>Answer the following questions:
    <ul>
        <li>What happens to the present value as we increase the compounding frequency? Why?</li>
        <li>What will happen if we increase the compounding frequency to infinity (or some extremely large number)?</li>
    </ul>
    <p style="text-align:right;">[15pts]</p>
    </div>
</div>

In [4]:
m_daily = 365

present_value_daily_compounding=CF/(1+rate/m_daily)**(m_daily*t)
print(present_value_daily_compounding)

#_#Computing present value with daily compounding using same CF, r, t. Created m_daily for daily compounding


90.4843615007843


*Your written response here*

-As we increase the compounding frequency the present value goes down. This is because the amount of "interest on interest"increases as we compound more frequently.

-If we increase the compoudning frequency to infinity it will continue to get smaller but the amount at which the present value decreases will decrease.

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 4</b>
    </div>
    <div class = "panel-body"> 
    Recall the formula for calculating present value under continuous compounding:
    <br><br>
    $$PV = CF * e^{-rt}$$
    <br>Using the same inputs thus far, calculate the cashflow's present value under continuous compounding.
    <p style="text-align:right;">[15pts]</p>
    </div>
</div>

In [5]:
import numpy as np

#_#numpy contains external data such as exp, and Sqrt.

In [6]:
e_value=np.exp(-rate*t)
print(e_value)

#_#Computing present value with continious compounding using same CF, r, t.

0.9048374180359595


In [7]:
present_value_continuously_compounded=CF*e_value
print(present_value_continuously_compounded)

#_#Multiplying continuously compounded PV by the cashflow

90.48374180359595


<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 5</b>
    </div>
    <div class = "panel-body"> 
    Recall that the net present value of a stream of cashflows is simply the sum of the present values of each individual cashflow:
    <br><br>
    $$NPV = \sum_{t}PV(CF_t)$$
    <br>Calculate the net present value for the following cashflows at years 1, 2, and 5 of \$100, \$110, and \$200, respectively. Assume annual compounding at a rate of 2%.
    <p style="text-align:right;">[15pts]</p>
    </div>
</div>

In [8]:
CF_1 = 100
CF_2 = 110
CF_3 = 200

t1 = 1
t2 = 2
t3 = 5

#_#three cashflows at t(1,2,5) are defined

In [9]:
PV_1 = (CF_1)/(1+rate)**t1
PV_2 = (CF_2)/(1+rate)**t2
PV_3 = (CF_3)/(1+rate)**t3

print(PV_1)
print(PV_2)
print(PV_3)

#_#finding PV of each cashflow

95.23809523809524
99.77324263038548
156.70523329369178


In [10]:
PV_list = [PV_1,PV_2,PV_3]

net_present_value = np.sum(PV_list)
print(net_present_value)

#_#computing the sum of the present values

351.7165711621725


<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 6</b>
    </div>
    <div class = "panel-body"> 
    The price of a risk-free coupon bearing bond is simply the NPV of its future generated cashflow stream. But given the consistent periodic coupon payments, we can treat the coupons as an annuity and simplify the NPV formula: 
    <br><br>
    $$B = F \left(\frac{c}{m}\right)\frac{1 - \left(1 + \frac{r}{m}\right)^{-mT}}{\frac{r}{m}}  + F(1 + r)^{-T}$$
    <br>Where $B$ is the bond price, $F$ is the face amount, $m$ is the coupon payment frequency, $T$ is the maturity, and interest rate $r$.
    <br><br>Apply this formula to calculate the price of a bond with \$1000 face value, maturing in 30 years, paying a semi-annual coupon of 4.5%, assuming a constant risk-free interest rate of 2%.
    <br><br>Answer the following questions: 
    <ul>
        <li>Is this a premium or a discount bond?</li>
        <li>What might cause this bond to trade at a market price that is different from this calculated price?</li>
    </ul>
    <p style="text-align:right;">[20pts]</p>
    </div>
</div>

In [11]:
face_value = 1000
t = 30 
m = 2
rf = .02
coupon = .045

bond_price = face_value*(coupon/m)*((1-(1+rf/m)**(-m*t))/(rf/m))+face_value*(1+rf)**(-t)
print(bond_price)

#_#Here we set the numbers in the task to a variable and set the bond price to the given equation with our assigned variables and printed the bond price.


1563.5592531199522


*Your written response here*
-This is a premium bond

-The rates could cause the bond price to change. If the risk-free rate were to be above the coupon rate the bond would have been a discount bond



<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 7</b>
    </div>
    <div class = "panel-body"> 
    Recall that the dividend discount model is a stock valuation model that attempts to value a stock as if it was a bond, based on its dividends. It assumes that a company will continue to pay its dividend uninterrupted in perpituity, and that the stock price is simply the sum of all of its future dividend payments, discounted back to their present value:
    <br><br>
    $$P = \frac{D_1}{r - g}$$
    <br>Where $P$ is the stock price, $D$ is the company's annual dividend payment in the next year, and $r$ is the company's cost of capital, and $g$ is the expected dividend growth rate.
    <br><br>As of June 2020, AAPL paid an annual dividend \$3.28 with a most recent dividend growth rate of 6.49%, and an estimated cost of capital of 9%. According to the dividend discount model, what is the value of AAPL stock?
    <br><br>How does that compare with the market price of AAPL ($364.8 as of June 2020). What are some shortcomings of the dividend discount model? How would the dividend discount model value TSLA?
    <p style="text-align:right;">[20pts]</p>
    </div>
</div>

In [12]:
dividend_0 = 3.28
div_growth = .0649

dividend_1 = dividend_0*(1+div_growth)
print(dividend_1)

#_#we had to use dividend 0 to calculate dividend 1 by using the growth rate

3.4928719999999998


In [13]:
est_capital = .09

stock_price = dividend_1/(est_capital-div_growth)

print(stock_price)

#_#set stock price = to the given formula using our variables


139.1582470119522


*Your written response here*
139.16 is much lower than the market price of AAPl as of June 2020. 

Some shortcomings of this model is the limited number of variables it takes into account. (e.g. It could be better by taking the market volatility into account)

This model would value TSLA much higher than it will actually be a few years from now. TSLA cannot continue to rise at this rate and therefore this model would incorrectly value TSLA.
